File: examples/typescript/transaction/src/__tests__/transaction-logic.test.ts
import { describe, it, expect, beforeEach } from 'vitest';
import { TransactionLogic, EventBook } from '../transaction-logic.js';
import { TransactionState, emptyTransactionState, LineItem } from '../transaction-state.js';
import { CommandValidationError, StatusCode } from '../command-validation-error.js';

describe('TransactionLogic', () => {
  let logic: TransactionLogic;

  beforeEach(() => {
    logic = new TransactionLogic();
  });

  describe('rebuildState', () => {
    it('returns empty state for null event book', () => {
      const state = logic.rebuildState(null);

      expect(state.customerId).toBe('');
      expect(state.items).toEqual([]);
      expect(state.subtotalCents).toBe(0);
      expect(state.status).toBe('');
    });

    it('returns empty state for empty event book', () => {
      const eventBook: EventBook = { pages: [] };

      const state = logic.rebuildState(eventBook);

      expect(state.customerId).toBe('');
      expect(state.status).toBe('');
    });

    it('applies TransactionCreated event', () => {
      const eventBook: EventBook = {
        pages: [
          {
            typeUrl: 'type.examples/examples.TransactionCreated',
            data: {
              customerId: 'cust-001',
              items: [{ productId: 'SKU-001', name: 'Widget', quantity: 2, unitPriceCents: 1000 }],
              subtotalCents: 2000,
            },
          },
        ],
      };

      const state = logic.rebuildState(eventBook);

      expect(state.customerId).toBe('cust-001');
      expect(state.subtotalCents).toBe(2000);
      expect(state.status).toBe('pending');
    });

    it('applies DiscountApplied event', () => {
      const eventBook: EventBook = {
        pages: [
          {
            typeUrl: 'type.examples/examples.TransactionCreated',
            data: { customerId: 'cust-001', items: [], subtotalCents: 2000 },
          },
          {
            typeUrl: 'type.examples/examples.DiscountApplied',
            data: { discountType: 'percentage', discountCents: 200 },
          },
        ],
      };

      const state = logic.rebuildState(eventBook);

      expect(state.discountCents).toBe(200);
      expect(state.discountType).toBe('percentage');
    });

    it('applies TransactionCompleted event', () => {
      const eventBook: EventBook = {
        pages: [
          {
            typeUrl: 'type.examples/examples.TransactionCreated',
            data: { customerId: 'cust-001', items: [], subtotalCents: 2000 },
          },
          {
            typeUrl: 'type.examples/examples.TransactionCompleted',
            data: { finalTotalCents: 2000 },
          },
        ],
      };

      const state = logic.rebuildState(eventBook);

      expect(state.status).toBe('completed');
    });

    it('applies TransactionCancelled event', () => {
      const eventBook: EventBook = {
        pages: [
          {
            typeUrl: 'type.examples/examples.TransactionCreated',
            data: { customerId: 'cust-001', items: [], subtotalCents: 2000 },
          },
          {
            typeUrl: 'type.examples/examples.TransactionCancelled',
            data: { reason: 'customer request' },
          },
        ],
      };

      const state = logic.rebuildState(eventBook);

      expect(state.status).toBe('cancelled');
    });
  });

  describe('calculateSubtotal', () => {
    it('calculates subtotal from items', () => {
      const items: LineItem[] = [
        { productId: 'SKU-001', name: 'Widget', quantity: 2, unitPriceCents: 1000 },
        { productId: 'SKU-002', name: 'Gadget', quantity: 1, unitPriceCents: 2500 },
      ];

      const subtotal = logic.calculateSubtotal(items);

      expect(subtotal).toBe(4500);
    });

    it('returns 0 for empty items', () => {
      expect(logic.calculateSubtotal([])).toBe(0);
    });
  });

  describe('handleCreateTransaction', () => {
    it('returns TransactionCreated event for new transaction', () => {
      const state = emptyTransactionState();
      const items: LineItem[] = [
        { productId: 'SKU-001', name: 'Widget', quantity: 2, unitPriceCents: 1000 },
      ];

      const event = logic.handleCreateTransaction(state, 'cust-001', items);

      expect(event.customerId).toBe('cust-001');
      expect(event.subtotalCents).toBe(2000);
      expect(event.items).toEqual(items);
      expect(event.createdAt).toBeDefined();
    });

    it('calculates subtotal for multiple items', () => {
      const state = emptyTransactionState();
      const items: LineItem[] = [
        { productId: 'SKU-001', name: 'Widget', quantity: 2, unitPriceCents: 1000 },
        { productId: 'SKU-002', name: 'Gadget', quantity: 1, unitPriceCents: 2500 },
      ];

      const event = logic.handleCreateTransaction(state, 'cust-002', items);

      expect(event.subtotalCents).toBe(4500);
    });

    it('throws failed_precondition for existing transaction', () => {
      const state: TransactionState = {
        customerId: 'cust-001',
        items: [],
        subtotalCents: 2000,
        discountCents: 0,
        discountType: '',
        status: 'pending',
      };
      const items: LineItem[] = [
        { productId: 'SKU-001', name: 'Widget', quantity: 1, unitPriceCents: 1000 },
      ];

      expect(() => logic.handleCreateTransaction(state, 'cust-002', items)).toThrow(
        CommandValidationError
      );

      try {
        logic.handleCreateTransaction(state, 'cust-002', items);
      } catch (e) {
        expect((e as CommandValidationError).statusCode).toBe(StatusCode.FAILED_PRECONDITION);
      }
    });

    it('throws invalid_argument for empty customer ID', () => {
      const state = emptyTransactionState();
      const items: LineItem[] = [
        { productId: 'SKU-001', name: 'Widget', quantity: 1, unitPriceCents: 1000 },
      ];

      expect(() => logic.handleCreateTransaction(state, '', items)).toThrow(
        CommandValidationError
      );

      try {
        logic.handleCreateTransaction(state, '', items);
      } catch (e) {
        expect((e as CommandValidationError).statusCode).toBe(StatusCode.INVALID_ARGUMENT);
      }
    });

    it('throws invalid_argument for empty items', () => {
      const state = emptyTransactionState();

      expect(() => logic.handleCreateTransaction(state, 'cust-001', [])).toThrow(
        CommandValidationError
      );

      try {
        logic.handleCreateTransaction(state, 'cust-001', []);
      } catch (e) {
        expect((e as CommandValidationError).statusCode).toBe(StatusCode.INVALID_ARGUMENT);
      }
    });
  });

  describe('handleApplyDiscount', () => {
    const pendingState: TransactionState = {
      customerId: 'cust-001',
      items: [],
      subtotalCents: 2000,
      discountCents: 0,
      discountType: '',
      status: 'pending',
    };

    it('applies percentage discount correctly', () => {
      const event = logic.handleApplyDiscount(pendingState, 'percentage', 10);

      expect(event.discountType).toBe('percentage');
      expect(event.value).toBe(10);
      expect(event.discountCents).toBe(200); // 10% of 2000
    });

    it('applies fixed discount correctly', () => {
      const event = logic.handleApplyDiscount(pendingState, 'fixed', 500);

      expect(event.discountType).toBe('fixed');
      expect(event.discountCents).toBe(500);
    });

    it('includes coupon code when provided', () => {
      const event = logic.handleApplyDiscount(pendingState, 'percentage', 10, 'SAVE10');

      expect(event.couponCode).toBe('SAVE10');
    });

    it('throws failed_precondition for non-pending transaction', () => {
      const completedState: TransactionState = {
        ...pendingState,
        status: 'completed',
      };

      expect(() => logic.handleApplyDiscount(completedState, 'percentage', 10)).toThrow(
        CommandValidationError
      );

      try {
        logic.handleApplyDiscount(completedState, 'percentage', 10);
      } catch (e) {
        expect((e as CommandValidationError).statusCode).toBe(StatusCode.FAILED_PRECONDITION);
      }
    });
  });

  describe('handleCompleteTransaction', () => {
    const pendingState: TransactionState = {
      customerId: 'cust-001',
      items: [],
      subtotalCents: 2000,
      discountCents: 0,
      discountType: '',
      status: 'pending',
    };

    it('returns TransactionCompleted event', () => {
      const event = logic.handleCompleteTransaction(pendingState, 'card');

      expect(event.finalTotalCents).toBe(2000);
      expect(event.paymentMethod).toBe('card');
      expect(event.loyaltyPointsEarned).toBe(20); // 1 point per dollar
      expect(event.completedAt).toBeDefined();
    });

    it('calculates final total with discount', () => {
      const stateWithDiscount: TransactionState = {
        ...pendingState,
        discountCents: 200,
      };

      const event = logic.handleCompleteTransaction(stateWithDiscount, 'cash');

      expect(event.finalTotalCents).toBe(1800);
      expect(event.loyaltyPointsEarned).toBe(18);
    });

    it('throws failed_precondition for non-pending transaction', () => {
      const emptyState = emptyTransactionState();

      expect(() => logic.handleCompleteTransaction(emptyState, 'card')).toThrow(
        CommandValidationError
      );

      try {
        logic.handleCompleteTransaction(emptyState, 'card');
      } catch (e) {
        expect((e as CommandValidationError).statusCode).toBe(StatusCode.FAILED_PRECONDITION);
      }
    });
  });

  describe('handleCancelTransaction', () => {
    const pendingState: TransactionState = {
      customerId: 'cust-001',
      items: [],
      subtotalCents: 2000,
      discountCents: 0,
      discountType: '',
      status: 'pending',
    };

    it('returns TransactionCancelled event', () => {
      const event = logic.handleCancelTransaction(pendingState, 'customer request');

      expect(event.reason).toBe('customer request');
      expect(event.cancelledAt).toBeDefined();
    });

    it('throws failed_precondition for completed transaction', () => {
      const completedState: TransactionState = {
        ...pendingState,
        status: 'completed',
      };

      expect(() => logic.handleCancelTransaction(completedState, 'too late')).toThrow(
        CommandValidationError
      );

      try {
        logic.handleCancelTransaction(completedState, 'too late');
      } catch (e) {
        expect((e as CommandValidationError).statusCode).toBe(StatusCode.FAILED_PRECONDITION);
      }
    });
  });
});

File: examples/go/transaction/logic/transaction_logic_test.go
package logic

import (
	"testing"

	"transaction/proto/angzarr"
	"transaction/proto/examples"

	"google.golang.org/protobuf/types/known/anypb"
)

func TestRebuildState_NilEventBook(t *testing.T) {
	logic := NewTransactionLogic()
	state := logic.RebuildState(nil)

	if state.Exists() {
		t.Error("expected empty state for nil event book")
	}
	if state.Status != "new" {
		t.Errorf("expected status 'new', got %q", state.Status)
	}
}

func TestRebuildState_EmptyEventBook(t *testing.T) {
	logic := NewTransactionLogic()
	state := logic.RebuildState(&angzarr.EventBook{})

	if state.Exists() {
		t.Error("expected empty state for empty event book")
	}
	if state.Status != "new" {
		t.Errorf("expected status 'new', got %q", state.Status)
	}
}

func TestRebuildState_WithTransactionCreated(t *testing.T) {
	logic := NewTransactionLogic()

	event := &examples.TransactionCreated{
		CustomerId:    "cust-001",
		SubtotalCents: 2000,
		Items: []*examples.LineItem{
			{ProductId: "SKU-001", Name: "Widget", Quantity: 2, UnitPriceCents: 1000},
		},
	}
	eventAny, _ := anypb.New(event)

	eventBook := &angzarr.EventBook{
		Pages: []*angzarr.EventPage{
			{Event: eventAny},
		},
	}

	state := logic.RebuildState(eventBook)

	if !state.Exists() {
		t.Error("expected state to exist")
	}
	if state.CustomerID != "cust-001" {
		t.Errorf("expected customer_id %q, got %q", "cust-001", state.CustomerID)
	}
	if state.SubtotalCents != 2000 {
		t.Errorf("expected subtotal_cents %d, got %d", 2000, state.SubtotalCents)
	}
	if state.Status != "pending" {
		t.Errorf("expected status 'pending', got %q", state.Status)
	}
}

func TestRebuildState_WithDiscountApplied(t *testing.T) {
	logic := NewTransactionLogic()

	created := &examples.TransactionCreated{
		CustomerId:    "cust-001",
		SubtotalCents: 2000,
	}
	createdAny, _ := anypb.New(created)

	discount := &examples.DiscountApplied{
		DiscountType:  "percentage",
		DiscountCents: 200,
	}
	discountAny, _ := anypb.New(discount)

	eventBook := &angzarr.EventBook{
		Pages: []*angzarr.EventPage{
			{Event: createdAny},
			{Event: discountAny},
		},
	}

	state := logic.RebuildState(eventBook)

	if state.DiscountCents != 200 {
		t.Errorf("expected discount_cents %d, got %d", 200, state.DiscountCents)
	}
	if state.DiscountType != "percentage" {
		t.Errorf("expected discount_type %q, got %q", "percentage", state.DiscountType)
	}
}

func TestRebuildState_WithTransactionCompleted(t *testing.T) {
	logic := NewTransactionLogic()

	created := &examples.TransactionCreated{
		CustomerId:    "cust-001",
		SubtotalCents: 2000,
	}
	createdAny, _ := anypb.New(created)

	completed := &examples.TransactionCompleted{
		FinalTotalCents: 2000,
		PaymentMethod:   "card",
	}
	completedAny, _ := anypb.New(completed)

	eventBook := &angzarr.EventBook{
		Pages: []*angzarr.EventPage{
			{Event: createdAny},
			{Event: completedAny},
		},
	}

	state := logic.RebuildState(eventBook)

	if state.Status != "completed" {
		t.Errorf("expected status 'completed', got %q", state.Status)
	}
}

func TestRebuildState_WithTransactionCancelled(t *testing.T) {
	logic := NewTransactionLogic()

	created := &examples.TransactionCreated{
		CustomerId:    "cust-001",
		SubtotalCents: 2000,
	}
	createdAny, _ := anypb.New(created)

	cancelled := &examples.TransactionCancelled{
		Reason: "customer request",
	}
	cancelledAny, _ := anypb.New(cancelled)

	eventBook := &angzarr.EventBook{
		Pages: []*angzarr.EventPage{
			{Event: createdAny},
			{Event: cancelledAny},
		},
	}

	state := logic.RebuildState(eventBook)

	if state.Status != "cancelled" {
		t.Errorf("expected status 'cancelled', got %q", state.Status)
	}
}

func TestHandleCreateTransaction_Success(t *testing.T) {
	logic := NewTransactionLogic()
	state := EmptyState()

	items := []*examples.LineItem{
		{ProductId: "SKU-001", Name: "Widget", Quantity: 2, UnitPriceCents: 1000},
	}

	event, err := logic.HandleCreateTransaction(state, "cust-001", items)

	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if event.CustomerId != "cust-001" {
		t.Errorf("expected customer_id %q, got %q", "cust-001", event.CustomerId)
	}
	if event.SubtotalCents != 2000 {
		t.Errorf("expected subtotal_cents %d, got %d", 2000, event.SubtotalCents)
	}
}

func TestHandleCreateTransaction_MultipleItems(t *testing.T) {
	logic := NewTransactionLogic()
	state := EmptyState()

	items := []*examples.LineItem{
		{ProductId: "SKU-001", Name: "Widget", Quantity: 2, UnitPriceCents: 1000},
		{ProductId: "SKU-002", Name: "Gadget", Quantity: 1, UnitPriceCents: 2500},
	}

	event, err := logic.HandleCreateTransaction(state, "cust-001", items)

	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if event.SubtotalCents != 4500 {
		t.Errorf("expected subtotal_cents %d, got %d", 4500, event.SubtotalCents)
	}
}

func TestHandleCreateTransaction_AlreadyExists(t *testing.T) {
	logic := NewTransactionLogic()
	state := &TransactionState{Status: "pending"}

	items := []*examples.LineItem{
		{ProductId: "SKU-001", Name: "Widget", Quantity: 1, UnitPriceCents: 1000},
	}

	_, err := logic.HandleCreateTransaction(state, "cust-001", items)

	if err == nil {
		t.Fatal("expected error for existing transaction")
	}
	cmdErr, ok := err.(*CommandError)
	if !ok {
		t.Fatalf("expected CommandError, got %T", err)
	}
	if cmdErr.Code != StatusFailedPrecondition {
		t.Errorf("expected FAILED_PRECONDITION, got %v", cmdErr.Code)
	}
}

func TestHandleCreateTransaction_EmptyCustomerID(t *testing.T) {
	logic := NewTransactionLogic()
	state := EmptyState()

	items := []*examples.LineItem{
		{ProductId: "SKU-001", Name: "Widget", Quantity: 1, UnitPriceCents: 1000},
	}

	_, err := logic.HandleCreateTransaction(state, "", items)

	if err == nil {
		t.Fatal("expected error for empty customer ID")
	}
	cmdErr, ok := err.(*CommandError)
	if !ok {
		t.Fatalf("expected CommandError, got %T", err)
	}
	if cmdErr.Code != StatusInvalidArgument {
		t.Errorf("expected INVALID_ARGUMENT, got %v", cmdErr.Code)
	}
}

func TestHandleCreateTransaction_NoItems(t *testing.T) {
	logic := NewTransactionLogic()
	state := EmptyState()

	_, err := logic.HandleCreateTransaction(state, "cust-001", nil)

	if err == nil {
		t.Fatal("expected error for no items")
	}
	cmdErr, ok := err.(*CommandError)
	if !ok {
		t.Fatalf("expected CommandError, got %T", err)
	}
	if cmdErr.Code != StatusInvalidArgument {
		t.Errorf("expected INVALID_ARGUMENT, got %v", cmdErr.Code)
	}
}

func TestHandleApplyDiscount_Percentage(t *testing.T) {
	logic := NewTransactionLogic()
	state := &TransactionState{
		Status:        "pending",
		SubtotalCents: 2000,
	}

	event, err := logic.HandleApplyDiscount(state, "percentage", 10, "")

	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if event.DiscountCents != 200 {
		t.Errorf("expected discount_cents %d, got %d", 200, event.DiscountCents)
	}
	if event.DiscountType != "percentage" {
		t.Errorf("expected discount_type %q, got %q", "percentage", event.DiscountType)
	}
}

func TestHandleApplyDiscount_Fixed(t *testing.T) {
	logic := NewTransactionLogic()
	state := &TransactionState{
		Status:        "pending",
		SubtotalCents: 2000,
	}

	event, err := logic.HandleApplyDiscount(state, "fixed", 500, "")

	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if event.DiscountCents != 500 {
		t.Errorf("expected discount_cents %d, got %d", 500, event.DiscountCents)
	}
}

func TestHandleApplyDiscount_FixedCappedAtSubtotal(t *testing.T) {
	logic := NewTransactionLogic()
	state := &TransactionState{
		Status:        "pending",
		SubtotalCents: 300,
	}

	event, err := logic.HandleApplyDiscount(state, "fixed", 500, "")

	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if event.DiscountCents != 300 {
		t.Errorf("expected discount_cents %d (capped at subtotal), got %d", 300, event.DiscountCents)
	}
}

func TestHandleApplyDiscount_Coupon(t *testing.T) {
	logic := NewTransactionLogic()
	state := &TransactionState{
		Status:        "pending",
		SubtotalCents: 2000,
	}

	event, err := logic.HandleApplyDiscount(state, "coupon", 0, "SAVE5")

	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if event.DiscountCents != 500 {
		t.Errorf("expected discount_cents %d, got %d", 500, event.DiscountCents)
	}
	if event.CouponCode != "SAVE5" {
		t.Errorf("expected coupon_code %q, got %q", "SAVE5", event.CouponCode)
	}
}

func TestHandleApplyDiscount_NotPending(t *testing.T) {
	logic := NewTransactionLogic()
	state := &TransactionState{
		Status:        "completed",
		SubtotalCents: 2000,
	}

	_, err := logic.HandleApplyDiscount(state, "percentage", 10, "")

	if err == nil {
		t.Fatal("expected error for non-pending transaction")
	}
	cmdErr := err.(*CommandError)
	if cmdErr.Code != StatusFailedPrecondition {
		t.Errorf("expected FAILED_PRECONDITION, got %v", cmdErr.Code)
	}
}

func TestHandleApplyDiscount_InvalidPercentage(t *testing.T) {
	logic := NewTransactionLogic()
	state := &TransactionState{
		Status:        "pending",
		SubtotalCents: 2000,
	}

	_, err := logic.HandleApplyDiscount(state, "percentage", 150, "")

	if err == nil {
		t.Fatal("expected error for invalid percentage")
	}
	cmdErr := err.(*CommandError)
	if cmdErr.Code != StatusInvalidArgument {
		t.Errorf("expected INVALID_ARGUMENT, got %v", cmdErr.Code)
	}
}

func TestHandleApplyDiscount_UnknownType(t *testing.T) {
	logic := NewTransactionLogic()
	state := &TransactionState{
		Status:        "pending",
		SubtotalCents: 2000,
	}

	_, err := logic.HandleApplyDiscount(state, "unknown", 10, "")

	if err == nil {
		t.Fatal("expected error for unknown discount type")
	}
	cmdErr := err.(*CommandError)
	if cmdErr.Code != StatusInvalidArgument {
		t.Errorf("expected INVALID_ARGUMENT, got %v", cmdErr.Code)
	}
}

func TestHandleCompleteTransaction_Success(t *testing.T) {
	logic := NewTransactionLogic()
	state := &TransactionState{
		Status:        "pending",
		SubtotalCents: 2000,
		DiscountCents: 0,
	}

	event, err := logic.HandleCompleteTransaction(state, "card")

	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if event.FinalTotalCents != 2000 {
		t.Errorf("expected final_total_cents %d, got %d", 2000, event.FinalTotalCents)
	}
	if event.PaymentMethod != "card" {
		t.Errorf("expected payment_method %q, got %q", "card", event.PaymentMethod)
	}
	if event.LoyaltyPointsEarned != 20 {
		t.Errorf("expected loyalty_points_earned %d, got %d", 20, event.LoyaltyPointsEarned)
	}
}

func TestHandleCompleteTransaction_WithDiscount(t *testing.T) {
	logic := NewTransactionLogic()
	state := &TransactionState{
		Status:        "pending",
		SubtotalCents: 2000,
		DiscountCents: 200,
	}

	event, err := logic.HandleCompleteTransaction(state, "cash")

	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if event.FinalTotalCents != 1800 {
		t.Errorf("expected final_total_cents %d, got %d", 1800, event.FinalTotalCents)
	}
	if event.LoyaltyPointsEarned != 18 {
		t.Errorf("expected loyalty_points_earned %d, got %d", 18, event.LoyaltyPointsEarned)
	}
}

func TestHandleCompleteTransaction_NotPending(t *testing.T) {
	logic := NewTransactionLogic()
	state := EmptyState()

	_, err := logic.HandleCompleteTransaction(state, "card")

	if err == nil {
		t.Fatal("expected error for non-pending transaction")
	}
	cmdErr := err.(*CommandError)
	if cmdErr.Code != StatusFailedPrecondition {
		t.Errorf("expected FAILED_PRECONDITION, got %v", cmdErr.Code)
	}
}

func TestHandleCancelTransaction_Success(t *testing.T) {
	logic := NewTransactionLogic()
	state := &TransactionState{
		Status:        "pending",
		SubtotalCents: 2000,
	}

	event, err := logic.HandleCancelTransaction(state, "customer request")

	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if event.Reason != "customer request" {
		t.Errorf("expected reason %q, got %q", "customer request", event.Reason)
	}
}

func TestHandleCancelTransaction_NotPending(t *testing.T) {
	logic := NewTransactionLogic()
	state := &TransactionState{
		Status: "completed",
	}

	_, err := logic.HandleCancelTransaction(state, "too late")

	if err == nil {
		t.Fatal("expected error for non-pending transaction")
	}
	cmdErr := err.(*CommandError)
	if cmdErr.Code != StatusFailedPrecondition {
		t.Errorf("expected FAILED_PRECONDITION, got %v", cmdErr.Code)
	}
}

func TestNextSequence_NilEventBook(t *testing.T) {
	seq := NextSequence(nil)
	if seq != 0 {
		t.Errorf("expected sequence 0, got %d", seq)
	}
}

func TestNextSequence_EmptyEventBook(t *testing.T) {
	seq := NextSequence(&angzarr.EventBook{})
	if seq != 0 {
		t.Errorf("expected sequence 0, got %d", seq)
	}
}

func TestNextSequence_WithEvents(t *testing.T) {
	eventBook := &angzarr.EventBook{
		Pages: []*angzarr.EventPage{
			{},
			{},
			{},
		},
	}
	seq := NextSequence(eventBook)
	if seq != 3 {
		t.Errorf("expected sequence 3, got %d", seq)
	}
}
